test/server.test.js

```diff
@@ -2820,3 +2820,128 @@ test('async handler should discard value', function(t) {
         t.end();
     });
 });
+
+test('Server returns 400 on invalid method', function(t) {
+    SERVER.get('/snickers/bar', function echoId(req, res, next) {
+        res.send();
+        next();
+    });
+
+    var opts = {
+        hostname: '127.0.0.1',
+        port: PORT,
+        path: '/snickers/bar',
+        method: 'CANDYBARS',
+        agent: false
+    };
+    http.request(opts, function(res) {
+        t.equal(res.statusCode, 400);
+        t.equal(res.statusMessage, 'Bad Request');
+        res.on('data', function() {
+            t.fail('Data was sent on 400 error');
+        });
+        res.on('end', function() {
+            t.end();
+        });
+    }).end();
+});
+
+test('Server returns 431 when header size is too large', function(t) {
+    SERVER.get('/jellybeans', function echoId(req, res, next) {
+        res.send();
+        next();
+    });
+
+    var opts = {
+        hostname: '127.0.0.1',
+        port: PORT,
+        path: '/jellybeans',
+        method: 'GET',
+        agent: false,
+        headers: {
+            'jellybean-colors': 'purple,green,red,black,pink,'.repeat(1000)
+        }
+    };
+    http.request(opts, function(res) {
+        t.equal(res.statusCode, 431);
+        t.equal(res.statusMessage, 'Request Header Fields Too Large');
+        res.on('data', function() {
+            t.fail('Data was sent on 431 error');
+        });
+        res.on('end', function() {
+            t.end();
+        });
+    }).end();
+});
+
+test('Server supports adding custom clientError listener', function(t) {
+    SERVER.get('/popcorn', function echoId(req, res, next) {
+        res.send();
+        next();
+    });
+
+    SERVER.on('clientError', function(err, socket) {
+        if (err.code !== 'HPE_HEADER_OVERFLOW') {
+            t.fail('Expected HPE_HEADER_OVERFLOW but err.code was ' + err.code);
+        }
+        socket.write("HTTP/1.1 418 I'm a teapot\r\nConnection: close\r\n\r\n");
+        socket.destroy(err);
+    });
+
+    var opts = {
+        hostname: '127.0.0.1',
+        port: PORT,
+        path: '/popcorn',
+        method: 'GET',
+        agent: false,
+        headers: {
+            'jellybean-colors': 'purple,green,red,black,pink,'.repeat(1000)
+        }
+    };
+    http.request(opts, function(res) {
+        t.equal(res.statusCode, 418);
+        t.equal(res.statusMessage, "I'm a teapot");
+        res.on('data', function() {});
+        res.on('end', function() {
+            t.end();
+        });
+    }).end();
+});
+
+test('Server correctly handles multiple clientError listeners', function(t) {
```

Comment:
Not sure about this. So it is an event emitter, so obviously two handlers is technically valid, but the use case here is almost certainly to be sending a response in some cases, which shifts the burden onto each handler to make sure it properly checks before sending the response and closing the socket.  If a user does this in multiple handlers they could get some pretty unexpected error results if they are not careful.

Could a better user experience from restify be to error or warn when they try to add a second handler?

Reply:
@wesleytodd, I agree generally. This is more of a contrived test rather than a valid usage example. I added this test to make sure that the proxying is done correctly / doesn't emit the proxied event multiple times if multiple listeners are added.

My goal with this change was to restore the ability to add a `clientError` listener after #1895 removed it, so am matching the original behavior here. I don't think folks would do this very often if at all. One actual use case might be to do the response in one handler and log in another. I'm not opposed to adding a warning but feels beyond the scope of this change.

Reply:
cool, yeah makes sense. I just wanted to bring up this as I have thought before that sometimes these kind of api's from node core have foot guns, and I think frameworks are generally good places to shore up those gaps.  I for sure don't think this one is a common issue worth spending much time on. So 👍 to the change!